argon2_out_of_process_support.go

```diff
@@ -0,0 +1,990 @@
+
+	// Argon2OutOfProcessErrorInvalidThreads means that an invalid number of threads was supplied.
+	Argon2OutOfProcessErrorInvalidThreads Argon2OutOfProcessErrorType = "invalid-threads"
+
+	// Argon2OutOfProcessErrorUnexpectedInput means that there was an error with the combination
+	// of inputs associated with the supplied request.
+	Argon2OutOfProcessErrorUnexpectedInput Argon2OutOfProcessErrorType = "unexpected-input"
+
+	// Argon2OutOfProcessErrorTimeout means that the specified command timeout expired before
+	// the request was given a chance to start.
+	Argon2OutOfProcessErrorKDFTimeout Argon2OutOfProcessErrorType = "timeout-error"
+
+	// Argon2OutOfProcessErrorRestartProcess means that this process has already processed a
+	// good KDF request, and the process should exit and be replaced by a new one.
+	Argon2OutOfProcessErrorRestartProcess Argon2OutOfProcessErrorType = "restart-process"
+
+	// Argon2OutOfProcessErrorKDFUnexpected means that an unexpected error occurred when
+	// running the actual KDF operation.
+	Argon2OutOfProcessErrorKDFUnexpected Argon2OutOfProcessErrorType = "unexpected-kdf-error"
+
+	// Argon2OutOfProcessErrorUnexpected means that an unexpected error occurred without
+	// a more specific error type.
+	Argon2OutOfProcessErrorUnexpected Argon2OutOfProcessErrorType = "unexpected-error"
+)
+
+// Argon2OutOfProcessResponse is the response to a request for an argon2
+// operation in a remote process.
+type Argon2OutOfProcessResponse struct {
+	Command          Argon2OutOfProcessCommand   `json:"command"`                     // The input command
+	Key              []byte                      `json:"key,omitempty"`               // The derived key, if the input command was "derive"
+	Duration         time.Duration               `json:"duration,omitempty"`          // The duration, if the input command was "duration"
+	WatchdogResponse []byte                      `json:"watchdog-response,omitempty"` // The response to a watchdog ping, if the input command was "watchdog"
+	ErrorType        Argon2OutOfProcessErrorType `json:"error-type,omitempty"`        // The error type, if an error occurred
+	ErrorString      string                      `json:"error-string,omitempty"`      // The error string, if an error occurred
+}
+
+// Argon2OutOfProcessError is returned from [Argon2OutOfProcessResponse.Err]
+// if the response indicates an error, or directly from methods of the [Argon2KDF]
+// implementation created by [NewOutOfProcessArgon2KDF] when the received response indicates
+// that an error ocurred.
+type Argon2OutOfProcessError struct {
+	ErrorType   Argon2OutOfProcessErrorType
+	ErrorString string
+}
+
+// Error implements the error interface.
+func (e *Argon2OutOfProcessError) Error() string {
+	str := "cannot process request: " + string(e.ErrorType)
+	if e.ErrorString != "" {
+		str += " ("
+		str += e.ErrorString
+		str += ")"
+	}
```

Comment:
you can use strings.Builder(), it's available from 1.10+

Reply:
I'm using that now.